fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -1097,12 +1115,12 @@ private boolean shouldBuildIndex(@Nonnull IndexState indexState, @Nonnull IndexS
         }
     }
 
-    private boolean shouldNotClearExistingIndexEntries(@Nonnull IndexState indexState,
-                                                       @Nonnull IndexStatePrecondition indexStatePrecondition) {
+    private boolean shouldClearExistingIndexEntries(@Nonnull IndexState indexState,
+                                                    @Nonnull IndexStatePrecondition indexStatePrecondition) {
         // If the index state is DISABLED, it is expected that there is no existing index entry. But we would like
         // to clear it anyway to play safe.
-        return indexState == IndexState.WRITE_ONLY &&
-               indexStatePrecondition == IndexStatePrecondition.BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY;
+        return !(indexState != IndexState.WRITE_ONLY &&
+                 indexStatePrecondition != IndexStatePrecondition.BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY);
```

Comment:
Is this right?

I think this simplifies to:

```java
indexState == IndexState.WRITE_ONLY 
    || indexStatePrecondition == IndexStatePrecondition.BUILD_IF_DISABLED_CONTINUE_IF_WRITE_ONLY
```

So I think this will cause it to clear whenever the index is initially `WRITE_ONLY`, which I don't think was the intention.


I think you want:

```java
indexState != IndexState.WRITE_ONLY 
    || indexStatePrecondition != IndexStatePrecondition.BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY
```

Or perhaps:

```java
indexState != IndexState.WRITE_ONLY
    || (indexStatePrecodition != IndexStatePrecondition.BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY 
         && indexStatePrecondition != IndexStatePrecondition.ERROR_IF_DISABLED_CONTINUE_IF_WRITE_ONLY) 
```

(One could also imagine adding a `.continueIfWriteOnly()` boolean method to `IndexStatePrecondition` to clean up some of that. I think that was discussed earlier, and I think we might have rejected it?)

Reply:
Good catch! Yeah, my intent was
```
!(indexState == IndexState.WRITE_ONLY &&
                 indexStatePrecondition == IndexStatePrecondition.BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY);
```
which I feel more readable than the simplified one. But I somehow screwed it up.


Interestingly I did remember adding `.continueIfWriteOnly()` being discussed. I probably mistook it with other ideas. I think it sound good. Updated.